### PR TITLE
CPS-0013? | Better builtin data structures in Plutus

### PR DESCRIPTION
**Pull request by michaelpj**
Based on some discussions with Pi and others.

Some tasks today are difficult because we don't have the data structures to do them efficiently. This CPS outlines a couple of examples and gestures towards some possible solutions (which will require a proper CIP).

-----------------------------------------
[Rendered Version](https://github.com/michaelpj/CIPs/blob/master/CIP-%3F%3F%3F%3F/README.md)

### CODE REVIEW

File: CIP-????/README.md
Comment (rphair) [inline CIP-????/README.md:2]: The containing folder is currently `CIP-????` so can you please rename it to `CPS-`something... preferably something unique e.g. `data-structures` since this is from your `master` branch & you might want to submit another one?

File: CIP-????/README.md
Comment (rphair) [inline CIP-????/README.md:3]:
```suggestion
Better builtin datastructures in Plutus
```
The current PR title sounds both broad and specific enough to cover this: if not this one, we would need a final title of around this length.  Your stating the "problem" in terms of what we are hoping to achieve ("Better XXX") I think works well here.

File: CIP-????/README.md
Comment (rphair) [inline CIP-????/README.md:97]:
```suggestion
```
We don't have a section for this in the CPS template... to me the discussion at this link seems split between _Problem_ and _Use cases_, so could it be linked in one or both those places as appropriate?

**Review by rphair — COMMENTED** [review]
Very nice to see such a proposal sourced in cooperation with the community 🤓

**Comment by Quantumplation** [comment]
Thanks for taking the time to compile this @michaelpj!

Not sure if this is appropriate for a CPS, but in the spirit of inspiring some CIPs to solve this CPS, let me document a few of the rough ideas we kicked around on that twitter thread:

# Length-prefixed arrays with constant time lookup

New builtin type `array`, with a `length` and `type`, such as:
```
(con (array 1 integer) [1])
(con (array 3 string) ["1", "2", "3"])
(con (array 7 data) [I 1, B aabb, I 2, I 3, B ddee, B ffgg, I 7])
```

New builtin function `arrayLength` that returns the length of the array

New builtin function `newArray` that constructs a new array of a given length
```
(builtin.newArray 5 data)

evalutes to `(con (array 5 data) [bottom, bottom, bottom, bottom, bottom])
```
(or perhaps you provide the initial value?)

New builtin function `index`, which retreives the `nth` item, or faults if its out of bounds
```
(builtin.index 3 (con (array 7 data) [I 1, B aabb, I 2, I 3, B ddee, B ffaa, I 7]))

evalutes to `I 3`
```

New builtin function `set` that returns a new array, with the `nth` item replaced by a new value, or faults if its out of bounds
```
(builtin.set 3 (B beef) (con (array 7 data) [I 1, B aabb, I 2, I 3, B ddee, B ffgg, I 7]))

evalutes to (con (array 7 data) [I 1, B aabb, I 2, B beef, B ddee, B ffgg, I 7])
```

## Why is this useful? 
There are several good candidates in the script context that are "index heavy" (most notably the inputs and outputs, but also withdrawals) that would benefit from being made into arrays. Even if not, a single linear pass to construct the array, and then constant time lookups after that, this would save quadratic work in many contracts.

GADTs from languages that compile to UPLC could compile structs to "Constructors with one field, which is an array of the logical fields of the struct", and then any field access is a `builtin.head` followed by a constant time access for the field we need.

# Associative Arrays

I have less fidelity here, so either a new builtin or a new Plutus Data type that got accelerated with a hashmap implementation; ideally, one that mapped directly onto CBOR maps for efficient encoding.

Would likely need builtins for looking up by key, and inserting (producing a new map with the entry added).

## Why is this useful?
Structs could compile down to maps, for log (with very small constants) time lookup of fields.
Datums / Redeemers could contain maps from address to value, or maps from txInput to action to take on that input, etc.

# Preserve the order of inputs from the transaction

This one is kind of an orthogonal / out of the box solution, but is highlighted by the struggles in this CPS; in particular, the best optimization is just to not do the work at all anymore, and a huge source of indexing slowdown is having to re-sort the transaction inputs in the first place.

File: CIP-????/README.md
Comment (michaelpj) [reply CIP-????/README.md:2]: um, I very much picked the wrong branch here, oops :facepalm: I guess I'll leave it for now :(

File: CIP-????/README.md
Comment (michaelpj) [reply CIP-????/README.md:3]: I'm not sure. I think the problem _is_ broader, I think it's jumping to conclusions a bit to structure it directly around the solution we're currently thinking of!

File: CIP-????/README.md
Comment (michaelpj) [reply CIP-????/README.md:97]: Maybe we should add a `References` section. It's pretty useful.

Comment (Ryun1) [reply CIP-????/README.md:97]: Good suggestion, see [Issue #731](https://github.com/cardano-foundation/CIPs/issues/731).

File: CIP-????/README.md
Comment (rphair) [reply CIP-????/README.md:3]: OK sounds like we're good to go with the broad, brief title (our CIP titles read more like newspaper headlines than academic papers)

File: CIP-????/README.md
Comment (rphair) [reply CIP-????/README.md:2]: @michaelpj no problem here for now... but if another CIP comes in on the `master` branch it will have to use a different subdirectory name 😅

File: CIP-????/README.md
Comment (rphair) [reply CIP-????/README.md:97]: ... specifically @michaelpj we are talking about restoring the removed `## References` section in https://github.com/cardano-foundation/CIPs/issues/731#issuecomment-1874421788 and with consensus will add it back in via https://github.com/cardano-foundation/CIPs/pull/730.

File: CIP-????/README.md
Comment (zliu41) [inline CIP-????/README.md:67]: > Since map operations are currently linear

To be precise, operations that should be logarithmic are currently linear (e.g., insertion, deletion, member), whereas operations should be linear (or nlogn) are currently quadratic (e.g., union)

File: CIP-????/README.md
Comment (zliu41) [inline CIP-????/README.md:90]: `Value` alone should make this useful. Also Aiken users would certainly answer "yes".

File: CIP-????/README.md
Comment (Quantumplation) [reply CIP-????/README.md:49]: For example, the ledger reorders transaction inputs and reference inputs from what is on-chain, sorting them by transaction ID and then by index.

And for many cases (such as processing DEX orders), the order you process the list in is important, as each one impacts how the next one is evaluated. So we provide a list in the redeemer specifying which order to process the inputs in, and thus iterating over the list in that order is quadratic (or near quadratic) time.

File: CIP-????/README.md
Comment (michaelpj) [reply CIP-????/README.md:52]: This is assuming we didn't change Data.

File: CIP-????/README.md
Comment (michaelpj) [reply CIP-????/README.md:90]: Yes, I think those two are compelling

File: CIP-????/README.md
Comment (colll78) [reply CIP-????/README.md:67]: If we had a list with O(1) index lookup then `valueOf` would be constant time in the majority of DApps in practice. This is because every robust DApp already explicitly prevents non-protocol related tokens from being stored in protocol script UTxOs, and in the vast majority of cases the validator will have access to the currency symbol of all the protocol tokens that it must check for. In all such cases you can just predetermine the indices for the relevant protocol tokens and then use `elemAt index valuePairList` instead of `valueOf`

**Comment by michaelpj** [comment]
@rphair I think this is ready for review

File: CIP-????/README.md
Comment (rphair) [reply CIP-????/README.md:102]: I'm preparing a PR that will document optional sections like this in the CPS template the way we added them for CIPs in #730.  We should have added it there, but by the time that discussion concluded I couldn't recall any CPS examples.  I'll have that PR for CIP-9999 submitted (and maybe merged) before this in final review, and in any case I believe a `## References` section by community consensus by editor unanimity is OK.

BUT @michaelpj @Quantumplation I don't think this link belongs just hanging here.  People not present in that Twitter discussion will have difficulty pulling the relevant elements out of it... so can you
- add some introduction about what is at this link?
- then consider moving it to _Problem_ or _Goals_ according to what this discussion most appropriately defines?

**Review by rphair — COMMENTED** [review]
@michaelpj  all looks generally in order to me.  Have added it to CIP agenda # 81 for Review this time, since at meeting # 79 it only got Triage.  I expect we'll be able to assign it a number at the meeting & I'm ready to green-check this soon after that's done... unless any major outcry or new unresolved issues.  cc @Quantumplation

To avoid confusion for reviewers, can you rename the containing directory to something beginning with `CPS-` and re-link to the latest file in your branch in your original comment up top?

Also we agreed `## References` is OK but I made a https://github.com/cardano-foundation/CIPs/pull/638#discussion_r1471403744 that this item currently looks out of place there & needs to be defined a bit more & maybe moved into a relevant main section.

**Note** this was supposed to be combined with https://github.com/cardano-foundation/CIPs/pull/638#discussion_r1471356114 but my Internet connection went boogey.

File: CIP-????/README.md
Comment (rphair) [reply CIP-????/README.md:102]: p.s. the official inclusion of _References_ in the CPS has been proposed in https://github.com/cardano-foundation/CIPs/pull/756.

**Review by Crypto2099 — APPROVED** [review]
This looks good and in order for a CPS to me. Thanks @rphair @michaelpj @Quantumplation for getting this together. Definitely worthy of a number at the next meeting at the very least!

**Review by Ryun1 — APPROVED** [review]
This appears to be a well written problem statement, with good discussions, happy to see a number applied then merged.

**Review by rphair — COMMENTED** [review]
I"m also very happy with it but will wait until it gets a number (very likely in couple days' time after meeting) in the usual cycle before I ✅ it for merge.

File: CIP-????/README.md
Comment (rphair) [inline CIP-????/README.md:2]:
```suggestion
CPS: 13
```

**Review by rphair — COMMENTED** [review]
@michaelpj this has passed editorial review but zooming ahead too quickly to merge now.  Please update the number here & in your branch (and correcting the document link in your OP please)... it's on `Last Check` now for next meeting agenda, if not merged by popular request & consensus sooner 🚀

I think https://github.com/cardano-foundation/CIPs/pull/638#discussion_r1471426353 still needs to be resolved although I believe that's a document quality & not a "standards" issue & therefore up to @michaelpj @Quantumplation about whether it needs to be fixed; so I won't let the merge be contingent upon it.
- TL;DR (already documented above) I think our body of CIPs shouldn't have `References` which are links of unknown significance to social discussions included without context: our CIPs are generally written at a higher calibre than that.

**Comment by colll78** [comment]
https://github.com/cardano-foundation/CIPs/pull/767#issuecomment-1977363621

Here are benchmarks that are relevant to the discussion of the benefits of indexed data-structures. The illustrate how incredibly expensive recursive lookups are currently.

**Pull request merged**
